lib/Conversion/HWToSystemC/HWToSystemC.cpp

```diff
@@ -0,0 +1,64 @@
+//===- HWToSystemC.cpp - HW To SystemC Conversion Pass --------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This is the main HW to SystemC Conversion Pass Implementation.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Conversion/HWToSystemC.h"
+#include "../PassDetail.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/SystemC/SystemCOps.h"
+#include "mlir/Dialect/ControlFlow/IR/ControlFlowOps.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/BuiltinDialect.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace mlir;
+using namespace circt;
+using namespace hw;
+
+//===----------------------------------------------------------------------===//
+// Conversion Infrastructure
+//===----------------------------------------------------------------------===//
+
+static void populateLegality(ConversionTarget &target) {
+  target.addIllegalDialect<HWDialect>();
+  target.addLegalDialect<mlir::BuiltinDialect>();
+  target.addLegalDialect<systemc::SystemCDialect>();
+}
+
+//===----------------------------------------------------------------------===//
+// HW to SystemC Conversion Pass
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct HWToSystemCPass : public ConvertHWToSystemCBase<HWToSystemCPass> {
+  void runOnOperation() override;
+};
+} // namespace
+
+/// Create a HW to SystemC dialects conversion pass.
+std::unique_ptr<OperationPass<ModuleOp>> circt::createConvertHWToSystemCPass() {
+  return std::make_unique<HWToSystemCPass>();
+}
+
+/// This is the main entrypoint for the HW to SystemC conversion pass.
+void HWToSystemCPass::runOnOperation() {
+  MLIRContext &context = getContext();
+  ModuleOp module = getOperation();
+
+  ConversionTarget target(context);
+  TypeConverter typeConverter;
+  RewritePatternSet patterns(&context);
+  populateLegality(target);
+
+  if (failed(applyFullConversion(module, target, std::move(patterns))))
+    signalPassFailure();
+}
```

Comment:
Is this doing anything right now? Are there any tests that can be written?

Reply:
No, it does not do anything right now. It will fail on every invocation. My goal was to keep this PR free of any dialect specific logic to make it easy to review and allow more focus on the added operations and transformations by creating a dedicated PR for every operation and the associated lowering transformation. For this PR I could only add an empty `systemc.ctor` operation to every `hw.module` which does not make a lot of sense on its own. So I planned to add it with the `systemc.module` lowering. I just had to add the `systemc.ctor` operation itself because otherwise no documentation would have been generated and the `circt-doc` target would have failed.